main/lsp/ErrorReporter.cc

```diff
@@ -0,0 +1,94 @@
+#include "main/lsp/ErrorReporter.h"
+#include "core/lsp/TypecheckEpochManager.h"
+#include "main/lsp/LSPMessage.h"
+#include "main/lsp/LSPOutput.h"
+#include "main/lsp/json_types.h"
+
+namespace sorbet::realmain::lsp {
+using namespace std;
+ErrorReporter::ErrorReporter(shared_ptr<const LSPConfiguration> config) : config(config) {}
+
+const vector<u4> &ErrorReporter::getDiagnosticEpochs() const {
+    return diagnosticEpochs;
+};
+
+const UnorderedSet<core::FileRef> &ErrorReporter::getFilesThatHaveErrors() const {
+    return filesThatHaveErrors;
+};
+
+void ErrorReporter::setMaxFileId(u4 id) {
+    ENFORCE(id >= diagnosticEpochs.size());
+    diagnosticEpochs.resize(id + 1, 0);
+};
+
+void ErrorReporter::pushDiagnostics(u4 epoch, core::FileRef file, vector<unique_ptr<core::Error>> errors,
+                                    std::unique_ptr<core::GlobalState> &gs) {
```

Comment:
I think we can make `gs` a `const` reference to enforce that it does not get mutated (changed). AFAIK, `pushDiagnostics` shouldn't need to make any changes to it; it just uses it to look at files.

```suggestion
                                    const core::GlobalState &gs) {
```

This means that callsites will need to pass `gs` differently. Currently, callsites are passing a `unique_ptr<core::GlobalState> &`, but we'll want a `const core::GlobalState &`. You can do that with the `*` operator:

```
er.pushDiagnostics(epoch, file, move(errors), *gs)
```

Also, you'd be able to change uses of `*gs` in this function to just `gs`.